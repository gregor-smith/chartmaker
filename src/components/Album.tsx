import React, { FC } from 'react'
import { css, cx } from 'emotion'
import { Album as AlbumDetails, formatAlbumTitle } from '../state'


type Props = {
    details: AlbumDetails | null
    sizeRem: number
}


const baseStyle = css({
    backgroundColor: 'white',
    margin: '0.15rem'
})


export const Album: FC<Props> = ({ details, sizeRem }) => {
    const style = cx(
        baseStyle,
        css({
            width: `${sizeRem}rem`,
            height: `${sizeRem}rem`
        })
    )
    const image = details === null
        ? null
        : <img src={details.image.smallURL} alt={formatAlbumTitle(details)}/>
    return (
        <div className={style}>
            {image}
        </div>
    )
}
