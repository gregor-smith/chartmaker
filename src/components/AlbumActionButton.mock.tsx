const mock: typeof import('@/components/AlbumActionButton') = {
    AlbumActionButton: props =>
        <div {...props} className='mock-album-action-button'/>
}


export const AlbumActionButton = mock.AlbumActionButton
