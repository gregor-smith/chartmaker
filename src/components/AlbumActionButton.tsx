import { h, FunctionComponent } from 'preact'
import { css } from 'emotion'


type Props = {
    onClick: () => void
    title: string
}


const style = css({
    background: 'none',
    border: 'none'
})


export const AlbumActionButton: FunctionComponent<Props> = props =>
    <button {...props} class={style}/>
