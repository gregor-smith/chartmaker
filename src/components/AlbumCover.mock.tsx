import { cx } from 'emotion'


const mock: typeof import('@/components/AlbumCover') = {
    AlbumCover: ({
        album,
        size,
        overlayClass,
        children,
        highlighted,
        ...props
    }) => {
        const json = JSON.stringify(album)
        const className = cx('mock-album-cover', overlayClass)
        return (
            <div {...props} className={className}>
                {`Album: ${json}`}
                {`Size: ${size}`}
                {`Highlighted: ${highlighted}`}
                {children}
            </div>
        )
    }
}


export const AlbumCover = mock.AlbumCover
