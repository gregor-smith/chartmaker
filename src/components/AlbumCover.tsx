import React, { FC, DragEvent } from 'react'
import { css, cx } from 'emotion'

import { TEXT_COLOUR, ALBUM_PADDING_SIZE, highlightBackgroundStyle } from '@/style'
import { Album } from '@/types'


export type AlbumCoverProps = {
    album: Album
    size: string
    onDragStart: (event: DragEvent<HTMLDivElement>) => void
    onDragOver?: (event: DragEvent<HTMLDivElement>) => void
    onDragEnter?: (event: DragEvent<HTMLDivElement>) => void
    onDrop?: (event: DragEvent<HTMLDivElement>) => void
    onMouseEnter?: () => void
    overlayClass?: string
    highlighted?: boolean
}


const baseContainerStyle = css({
    position: 'relative',
    background: TEXT_COLOUR,
    margin: ALBUM_PADDING_SIZE
})

const imageStyle = css({
    position: 'absolute',
    top: 0,
    left: 0,
    width: '100%',
    maxWidth: '100%',
    height: '100%',
    objectFit: 'contain'
})

// the overlay is given an additional non-emotion class so it can be used as a
// selector in the container's style
const overlaySelector = 'overlay'

const baseOverlayStyle = cx(
    overlaySelector,
    imageStyle,
    css({ zIndex: 10 })
)


export const AlbumCover: FC<AlbumCoverProps> = ({
    album,
    size,
    onDragStart,
    onDragEnter,
    onDragOver,
    onDrop,
    onMouseEnter,
    children,
    overlayClass,
    highlighted
}) => {
    const overlayStyle = cx(baseOverlayStyle, overlayClass)
    const containerStyle = cx(
        baseContainerStyle,
        css({
            width: size,
            height: size,
            [`:not(:hover) .${overlaySelector}`]: {
                display: 'none'
            }
        }),
        highlighted === false
            ? highlightBackgroundStyle
            : undefined
    )

    let image: JSX.Element | undefined
    if (!album.placeholder) {
        image = (
            // lazy loading causes problems with html2canvas so keep it default
            <img className={imageStyle}
                src={album.url}
                alt={album.name}
                crossOrigin='anonymous'/>
        )
    }

    return (
        <div className={containerStyle}
                draggable={!album.placeholder}
                onDragStart={onDragStart}
                onDragEnter={onDragEnter}
                onDragOver={onDragOver}
                onDrop={onDrop}
                onMouseEnter={onMouseEnter}
                title={album.placeholder ? undefined : album.name}>
            {image}
            <div className={overlayStyle}>
                {children}
            </div>
        </div>
    )
}
