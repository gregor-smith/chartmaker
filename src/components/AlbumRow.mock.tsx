const mock: typeof import('@/components/AlbumRow') = {
    AlbumRow: ({ children }) =>
        <div className='mock-album-row'>
            {children}
        </div>
}


export const AlbumRow = mock.AlbumRow
