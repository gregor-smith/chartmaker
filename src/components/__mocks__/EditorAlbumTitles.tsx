import type { FC } from 'react'

import type { EditorAlbumTitlesProps } from '@/components/EditorAlbumTitles'


export const EditorAlbumTitles: FC<EditorAlbumTitlesProps> = ({
    groups,
    highlighted
}) => {
    const json = JSON.stringify(groups)
    return (
        <div className='mock-editor-album-titles'>
            {`Groups: ${json}`}
            {`Highlighted: ${highlighted}`}
        </div>
    )
}
