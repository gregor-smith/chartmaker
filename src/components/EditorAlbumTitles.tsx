import type { FC } from 'react'

import type { NamedAlbum } from '@/types'
import type { DispatchProps } from '@/reducer'
import { AlbumTitlesContainer } from '@/components/AlbumTitlesContainer'
import { AlbumTitle } from '@/components/AlbumTitle'
import { RenameAlbumButton } from '@/components/RenameAlbumButton'
import { DeleteAlbumButton } from '@/components/DeleteAlbumButton'
import { AlbumTitleGroup } from '@/components/AlbumTitleGroup'


export type EditorAlbumTitlesProps = DispatchProps & {
    groups: NamedAlbum[][]
    highlighted: number | undefined
}


export const EditorAlbumTitles: FC<EditorAlbumTitlesProps> = ({
    dispatch,
    groups,
    highlighted
}) => {
    const elements = groups.map((group, index) => {
        const titles = group.map(({ id, name }) =>
            <AlbumTitle key={id}
                    dispatch={dispatch}
                    id={id}
                    name={name}
                    highlighted={highlighted}>
                <RenameAlbumButton dispatch={dispatch} id={id}/>
                <DeleteAlbumButton dispatch={dispatch} id={id}/>
            </AlbumTitle>
        )
        return (
            <AlbumTitleGroup key={index}>
                {titles}
            </AlbumTitleGroup>
        )
    })
    return (
        <AlbumTitlesContainer dispatch={dispatch}>
            {elements}
        </AlbumTitlesContainer>
    )
}
