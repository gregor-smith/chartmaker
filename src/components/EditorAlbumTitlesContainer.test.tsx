import { render } from 'react-dom'

import type { NamedAlbum } from '@/types'
import { EditorAlbumTitles } from '@/components/EditorAlbumTitles'

import { createTestNamedAlbums, ignore, RenderContainer } from '@/test-utils/utils'


jest.mock('@/components/EditorAlbumTitleGroup')
jest.mock('@/components/AlbumTitlesContainer')


const container = new RenderContainer()


test.each<[ NamedAlbum[][], number | undefined ]>([
    [
        [
            createTestNamedAlbums(2),
            createTestNamedAlbums(3, 3),
            createTestNamedAlbums(2, 6)
        ],
        1
    ],
    [
        [
            createTestNamedAlbums(5),
            createTestNamedAlbums(3, 6)
        ],
        undefined
    ],
])('renders container and rows', (groups, highlighted) => {
    render(
        <EditorAlbumTitles dispatch={ignore}
            groups={groups}
            highlighted={highlighted}/>,
        container.element
    )

    expect(container.element).toMatchSnapshot()
})
