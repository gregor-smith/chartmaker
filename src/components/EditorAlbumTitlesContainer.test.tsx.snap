// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders container and rows 1`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-album-titles-container"
  >
    <div
      class="mock-editor-album-title-group"
    >
      Albums: [{"id":1,"name":"Test album 1","url":"https://test.com/1"},{"id":2,"name":"Test album 2","url":"https://test.com/2"}]
      Highlighted: 1
    </div>
    <div
      class="mock-editor-album-title-group"
    >
      Albums: [{"id":3,"name":"Test album 3","url":"https://test.com/3"}]
      Highlighted: 1
    </div>
    <div
      class="mock-editor-album-title-group"
    >
      Albums: []
      Highlighted: 1
    </div>
  </div>
</div>
`;

exports[`renders container and rows 2`] = `
<div
  id="test-render-container"
>
  <div
    class="mock-album-titles-container"
  >
    <div
      class="mock-editor-album-title-group"
    >
      Albums: [{"id":1,"name":"Test album 1","url":"https://test.com/1"},{"id":2,"name":"Test album 2","url":"https://test.com/2"},{"id":3,"name":"Test album 3","url":"https://test.com/3"},{"id":4,"name":"Test album 4","url":"https://test.com/4"},{"id":5,"name":"Test album 5","url":"https://test.com/5"}]
      Highlighted: undefined
    </div>
    <div
      class="mock-editor-album-title-group"
    >
      Albums: []
      Highlighted: undefined
    </div>
  </div>
</div>
`;
